### PR TITLE
[GEOS-12140]: Better Cropping on raster mask filter, rescaling to target resolution before the crop

### PR DESCRIPTION
**Pull request by dromagnoli**
[GEOS-12140](https://osgeo-org.atlassian.net/browse/GEOS-12140) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geoserver&utm_campaign=badge_info)  

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

The PR will be merged when all the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)), there is a code committer review, and the checklist has been fulfilled.

### CODE REVIEW

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java [hunk 1]
```diff
@@ -142,17 +155,142 @@ static GridCoverage2D read(GridCoverage2DReader delegate, WrapperPolicy policy,
...
+                                rasterFilter.getEnvelopeInternal(), grid.getCoordinateReferenceSystem2D()));
+
+                // The underlying reader may have returned a coverage with a different resolution than the one
+                // requested. The requested gridGeometry may have been limited too, due to reaching the
+                // Max Oversampling Factor.
+                //
+                // This happens for example when the data resolution is bad and the map is heavily oversampled.
+                // We want to scale it to the requested map raster extent before cropping to the geometry.
+
+                grid = scaleToRequestedSize(grid, getRequestedMapRasterArea(), interpolation);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java]: grid could be null, but scaleToRequestedSize does not check it, while it's checked below at line 182. I would just exit right away if the grid happened to be null. Please add a test for this.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java [hunk 2]
```diff
@@ -142,17 +155,142 @@ static GridCoverage2D read(GridCoverage2DReader delegate, WrapperPolicy policy,
...
+    }
+
+    private static <T> T getProperty(Object object, String property, Class<T> type) {
+        if (!OwsUtils.has(object, property)) {
+            return null;
+        }
+        return OwsUtils.property(object, property, type);
+    }
+
+    private static GridCoverage2D scaleToRequestedSize(
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:249]: Hum... the grid could be covering only part of the requested area, isn't this scaling its size up to the full requested size? That would make sense if the grid was covering the whole requested size.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java [hunk 3]
```diff
@@ -142,17 +155,142 @@ static GridCoverage2D read(GridCoverage2DReader delegate, WrapperPolicy policy,
...
+    private static Rectangle getRequestedMapRasterArea() {
+        Request request = Dispatcher.REQUEST.get();
+        if (request == null
+                || request.getOperation() == null
+                || request.getOperation().getParameters() == null) {
+            return null;
+        }
+
+        for (Object parameter : request.getOperation().getParameters()) {
+            Rectangle mapExtent = getRequestedMapRasterArea(parameter);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java]: Hum... looking for width/height is quite WMS-specific, and at that point you could use the dispatcher more directly (getting a WMS request object). In fact, this code path is fully WMS-specific... and at that point I would spare the reflection and just grab the parsed KVP from REQUEST, width and height should be in there, already parsed to integer.

If you want to make it fully protocol-independent instead, what about using the grid geometry being requested... shouldn't it indicate the area of interest and the target resoltuion?

**Review by aaime — COMMENTED** [review]
Overall looks good, see comments below.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java (hunk 3 above)
Comment (dromagnoli) [reply src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java]: Ok. I will follow that approach in getting from REQUEST.

In terms of using the requested gridGeometry it won't work.
At the beginning I was using the requested gridGeometry but there are pieces of code (I should re-check where, I think somewhere in the ReadResolutionCalculator) that won't go over the configured max oversampling factor. So when there is a big oversampling (default threshold is around 30 if my memory serves me right), they get limited and the target resolution modified and you get back again pixellated results.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java (hunk 2 above)
Comment (dromagnoli) [reply src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:249]: Good catch... I think that I could apply a correcting factor that relates requested envelope to grid envelope so that we apply the same "percentage" to the raster size.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java [hunk 4]
```diff
@@ -142,17 +157,224 @@ static GridCoverage2D read(GridCoverage2DReader delegate, WrapperPolicy policy,
...
+        Rectangle requestedGridRange = requestedMapArea.gridRange;
+        int requestedWidth = (int) requestedGridRange.getWidth();
+        int requestedHeight = (int) requestedGridRange.getHeight();
+        Envelope coveredRequestedEnvelope = scalingEnvelope.intersection(coverageRequestedEnvelope);
+        // WMS 1.3.0 with geographic CRS (e.g. EPSG:4326): axis 0 = lat = vertical, axis 1 = lon = horizontal.
+        // Swap span axes so WIDTH pixels map to lon extent and HEIGHT pixels map to lat extent.
+        boolean northEast =
+                CRS.getAxisOrder(coverageRequestedEnvelope.getCoordinateReferenceSystem()) == CRS.AxisOrder.NORTH_EAST;
+        if (northEast) {
+            requestedWidth = getRequestedSize(
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:275]: This branch is tricky but has no tests covering it.

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java [hunk 5]
```diff
@@ -142,17 +157,224 @@ static GridCoverage2D read(GridCoverage2DReader delegate, WrapperPolicy policy,
...
+        if (coverageCRS == null) {
+            return null;
+        }
+
+        try {
+            CoordinateReferenceSystem requestedCRS = referencedRequestedEnvelope.getCoordinateReferenceSystem();
+            if (CRS.equalsIgnoreMetadata(requestedCRS, coverageCRS)) {
+                return referencedRequestedEnvelope;
+            }
+            return referencedRequestedEnvelope.transform(coverageCRS, true);
```
Comment (aaime) [inline src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:324]: Are there tests covering different CRSs, exercising the transformation?

**Review by aaime — CHANGES_REQUESTED** [review]
Had a second look, the code now looks good, but could use a bit more testing.
I would also like to see an end-to-end WMS GetMap test (as a rule of thumb, unit tests are great to cover corner cases, but some end-to-end testing should be implemented to ensure the oevrall fix persists over time).

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java (hunk 4 above)
Comment (dromagnoli) [reply src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:275]: Adding it

File: src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java (hunk 5 above)
Comment (dromagnoli) [reply src/main/src/main/java/org/geoserver/security/decorators/SecuredGridCoverage2DReader.java:324]: Adding it

**Comment by dromagnoli** [comment]
> Had a second look, the code now looks good, but could use a bit more testing. I would also like to see an end-to-end WMS GetMap test (as a rule of thumb, unit tests are great to cover corner cases, but some end-to-end testing should be implemented to ensure the oevrall fix persists over time).

Also added a getMap test involving a triangular crop on a low resolution image.
https://github.com/dromagnoli/geoserver/blob/63eaf83828e64e64141102b39a0bdba3270aa6a2/src/wms1_1/src/test/java/org/geoserver/wms/ResourceAccessManagerWMSTest.java#L390

**Comment by dromagnoli** [comment]
All feedbacks have been addressed.
